### ISSUE TITLE
combining files with different formats

### ISSUE THREAD

**Issue by arademaker** [dkmfbk/rdfpro#6]
```
rdfpro @r -w own-pt-fixed.nt wordnet-en-fixed.nt same-as.nt ../../wn30.ttl @w tudo-pro.ttl
```

The command above produced an invalid file. What should be the right syntax for combine ntriples file and a turtle file produzing a turtle one?

**Comment by fracor** [comment]
Hi and thanks for using RDFpro! :-)

The syntax is fine: mixing different formats in input is ok, and the output format is detected by the extension. So it is a problem of the tool.

Are you using the latest version of the tool (0.4.1)?

With the previous versions we had some problems with invalid turtle files produced by the turtle writer of Sesame (the library used by RDFpro) in presence of IRIs containing spaces or other particular characters (e.g., the '.' character).

If already using the latest version of RDFpro, it would be helpful if you could provide:
  * the output of RDFpro when executing the reported command 
  * the error message you get when trying to read the invalid file produced by RDFpro (you can read it with `rdfpro @r tudo-pro.ttl`).
  * the input files (or excerpts of them) that we can use to reproduce the error

**Comment by arademaker** [comment]
Yes. I am using 0.4.1 version

The output has no error message:

```
$ rdfpro @r -w tmp/ver-2/own-pt-fixed.nt tmp/ver-2/wordnet-en-fixed.nt tmp/ver-2/same-as.nt wn30.ttl @w tudo-pro.ttl
01:53:44(I) 4460649 triples read (274248 tr/s avg)
01:53:44(I) 4460649 triples written (274248 tr/s avg)
01:53:44(I) Done in 16 s
```

rapper tool doesn't complain either

```
$ rapper -c -i turtle tudo-pro.ttl
rapper: Parsing URI file:///Users/arademaker/work/openWordnet-PT/tudo-pro.ttl with parser turtle
rapper: Parsing returned 4460649 triples
````

which seems to be ok because ntriples is a subset of turtle, right?  But Hermit reasoner complained about this line

```
<https://w3id.org/own-pt/wn30-pt/instances/synset-01106808-n> <https://w3id.org/own-pt/wn30/schema/gloss> """o ato de transigir com ou entre grupos (exercendo actividades comerciais); \"nenhuma transação é possível sem ele\"; \"ele
     sempre foi honesto em suas transações comigo \"""" .
````

So, it can be actually a bug in the turtle parser of Hermit! It looks like a valid turtle. But after the creation of this issue I do get another problem with rdfpro+turtle

````
$ cat test.ttl
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix owl: <http://www.w3.org/2002/07/owl#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix wn30: <https://w3id.org/own-pt/wn30/schema/> .
@prefix nomlex: <https://w3id.org/own-pt/nomlex/schema/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix : <http://w3id.org/own-pt/example/> .

:ws a wn30:WordSense ;
   wn30:word :w1 ;
   wn30:word :w2 .

:w1 a wn30:Word .
:w2 a wn30:Word .

:w1 wn30:lexicalForm "w1 form"^^xsd:string ;
    wn30:lexicalForm "w2 form"^^xsd:string .
```

The output of rapper

```
$ rapper -i turtle -o rdfxml test.ttl
rapper: Parsing URI file:///Users/arademaker/work/openWordnet-PT/test.ttl with parser turtle
rapper: Serializing with serializer rdfxml
<?xml version="1.0" encoding="utf-8"?>
<rdf:RDF xmlns:nomlex="https://w3id.org/own-pt/nomlex/schema/" xmlns:owl="http://www.w3.org/2002/07/owl#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#" xmlns:wn30="https://w3id.org/own-pt/wn30/schema/" xmlns:xsd="http://www.w3.org/2001/XMLSchema#" xmlns="http://w3id.org/own-pt/example/">
  <rdf:Description rdf:about="http://w3id.org/own-pt/example/ws">
    <rdf:type rdf:resource="https://w3id.org/own-pt/wn30/schema/WordSense"/>
  </rdf:Description>
  <rdf:Description rdf:about="http://w3id.org/own-pt/example/ws">
    <wn30:word rdf:resource="http://w3id.org/own-pt/example/w1"/>
  </rdf:Description>
  <rdf:Description rdf:about="http://w3id.org/own-pt/example/ws">
    <wn30:word rdf:resource="http://w3id.org/own-pt/example/w2"/>
  </rdf:Description>
  <rdf:Description rdf:about="http://w3id.org/own-pt/example/w1">
    <rdf:type rdf:resource="https://w3id.org/own-pt/wn30/schema/Word"/>
  </rdf:Description>
  <rdf:Description rdf:about="http://w3id.org/own-pt/example/w2">
    <rdf:type rdf:resource="https://w3id.org/own-pt/wn30/schema/Word"/>
  </rdf:Description>
  <rdf:Description rdf:about="http://w3id.org/own-pt/example/w1">
    <wn30:lexicalForm rdf:datatype="http://www.w3.org/2001/XMLSchema#string">w1 form</wn30:lexicalForm>
  </rdf:Description>
  <rdf:Description rdf:about="http://w3id.org/own-pt/example/w1">
    <wn30:lexicalForm rdf:datatype="http://www.w3.org/2001/XMLSchema#string">w2 form</wn30:lexicalForm>
  </rdf:Description>
</rdf:RDF>
rapper: Parsing returned 7 triples
```

The command

```
rdfpro @r test.ttl wn30.ttl @w my.ttl
```

It skipped the datatype of the literals:

```
...
:w1 wn30:lexicalForm "w1 form" .
:w1 wn30:lexicalForm "w2 form" .
...
````

**Comment by fracor** [comment]
Concerning the first issue (invalid Turtle output when parsed with HermiT), the 'offending' triple

    <https://w3id.org/own-pt/wn30-pt/instances/synset-01106808-n> <https://w3id.org/own-pt/wn30/schema/gloss> """o ato de transigir com ou entre grupos (exercendo actividades comerciais); \"nenhuma transação é possível sem ele\"; \"ele
        sempre foi honesto em suas transações comigo \"""" .

seems perfectly legal to me, too. I tested with the latest version of HermiT that uses OWL API 3.4.3 for parsing RDF. HermiT was not able to read a file containing that triple. The problem seems to be the escaped `\"` next to the closing `"""` marker. This can be verified using a simple test file (valid according to original Turtle spec as well as RDF 1.1 Turtle spec):

    <ex:a> <ex:p> """a\""""

rdfpro, riot and rapper (the tools I tried) can read the file, while HermiT cannot (but if I remove the \" character Hermit is able to read the file). Interestingly, the same test file can be successfully read by my version of Protégé that uses the more recent OWL API 3.5.

Summing up, the problem is in the OWL API version used by HermiT. As a quick workaround, you can try using different output formats that might be better supported by the OWL API version used in HermiT (I would start with NTriples, as the test file above converted to NTriples was successfully parsed by HermiT).

Concerning the second issue (missing xsd:string datatype in rdfpro output), this is a result of rdfpro 0.4.1 being based on RDF 1.1 (due to the use of Sesame 2.8). In RDF 1.1, a literal has always a datatype [1]. If omitted in a particular syntax, the datatype is assumed to be xsd:string [1][2], therefore Sesame and thus rdfpro never emit the xsd:string datatype.

Is this a problem? If yes, we could investigate providing a 'legacy' RDF 1.0 mode in the next release of rdfpro (in the meanwhile, you could try using version 0.4 that still uses Sesame 2.7.x and thus is based on RDF 1.0).

[1] http://www.w3.org/TR/rdf11-concepts/#section-Graph-Literal section 3.3
[2] http://www.w3.org/TR/turtle/#turtle-literals section 2.5.1

**Comment by arademaker** [comment]
Thank you for your detail analysis of the problem. About:

> Concerning the second issue (missing xsd:string datatype in rdfpro output), this is a result of rdfpro 0.4.1 being based on RDF 1.1 (due to the use of Sesame 2.8). In RDF 1.1, a literal has always a datatype [1]. If omitted in a particular syntax, the datatype is assumed to be xsd:string [1][2], therefore Sesame and thus rdfpro never emit the xsd:string datatype.
> Is this a problem? If yes, we could investigate providing a 'legacy' RDF 1.0 mode in the next release of rdfpro (in the meanwhile, you could try using version 0.4 that still uses Sesame 2.7.x and thus is based on RDF 1.0).
> 

My personal point of view is that information should be preserved always if it is valid. I would not remove explicit information because of a default suggested behavior.

Best,
Alexandre

**Comment by fracor** [comment]
Unfortunately, the information about whether a literal was originally a plain literal or an `xsd:string` literal is no more available when writing the output file. rdfpro loses this information as soon as it parses a file, since accordingly to RDF 1.1 it normalizes plain literals to `xsd:string` literals for internal processing (which can be more complex than simple file merging).

Keeping track of the original datatype (or lack of datatype) before its normalization and using this information when writing output data is possible but not enough for proper compatibility with RDF 1.0 applications, as RDF 1.1 rules would still require rdfpro to treat `"x"` and `"x"^^xsd:string` as the same literal internally, possibly causing different triples under RDF 1.0 rules to be merged because equal under RDF 1.1 rules.

For a better interplay between RDFpro and RDF 1.0 tools, I think the cleanest solution is to introduce an RDF 1.0 legacy mode where no datatype normalization occurs.

I'll close this issue and open an enhancement request so to keep track of this aspect in future development.

Thanks for your feedback!
